### PR TITLE
channel snapshot_solo transaction

### PR DESCRIPTION
**Pull request by velzevur**
PT: [Add single-signed state channel snapshot tx](https://www.pivotaltracker.com/story/show/158759996)

Most notable thing would be the proposal of the refactoring in aesc_tx_SUITE

### CODE REVIEW

**Comment by velzevur** [comment]
Left to do 

- [ ] Bump DB version

- [ ]  Add a line in release notes 

- [ ] Make more happy path tests to make sure a snapshot can be posted after a withdrawal/deposit/snapshot

**Comment by lucafavatella** [comment]
I see corresponding protocol PR https://github.com/aeternity/protocol/pull/129 is merged.

**Comment by velzevur** [comment]
@lucafavatella 
> I see corresponding protocol PR aeternity/protocol#129 is merged.
No. The snapshot is a distinct new transaction type and has nothing to do with solo close (and aeternity/protocol#129 for that matter). Corresponding protocol PR is cooking

**Comment by velzevur** [comment]
Only protocol repo changes are pending but those are not in the scope of this PR. Removing the WIP label

File: apps/aechannel/src/aesc_snapshot_solo_tx.erl [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+%%%===================================================================
+
+-spec new(map()) -> {ok, aetx:tx()}.
+new(#{channel_id := ChannelIdBin,
+      from       := FromPubKey,
+      payload    := Payload,
+      fee        := Fee,
+      nonce      := Nonce} = Args) ->
+    Tx = #channel_snapshot_solo_tx{
+            channel_id = aec_id:create(channel, ChannelIdBin),
```
Comment (gorillainduction) [inline apps/aechannel/src/aesc_snapshot_solo_tx.erl]: This doesn't match the API where `aec_id:id()` is passed into the transaction.

Please rebase on master and amend this.

**Comment by gorillainduction** [comment]
Missing:
* Creation of `snapshot_solo_tx` through the API.
* Tests of the above API in `aehttp_integration_SUITE`

Other than that, it looks good.

Very good refactoring of `aesc_fsm_SUITE.erl`

**Comment by velzevur** [comment]
Addressed the aec_id comment, left to address the endpoint one

**Comment by velzevur** [comment]
Added the missing convience endpoint with corresponding tests

File: apps/aechannel/src/aesc_withdraw_tx.erl [hunk 2]
```diff
@@ -282,7 +282,9 @@ check_channel(ChannelId, Amount, ToPubKey, Round, Trees) ->
 -spec check_amount(aesc_channels:channel(), aesc_channels:amount()) ->
                           ok | {error, not_enough_channel_funds}.
 check_amount(Channel, Amount) ->
-    case aesc_channels:total_amount(Channel) >= Amount of
+    MaxWithdrawableAmt = aesc_channels:total_amount(Channel) -
+                         2 * aesc_channels:channel_reserve(Channel),
```
[the remark this answers is not in the document]
Comment (velzevur) [reply apps/aechannel/src/aesc_withdraw_tx.erl]: Both participants are supposed to keep at least `channel_reserve` in their channel balance. We can not enforce it but if the total balance goes less than what is supposed to be - this is surely something we can enforce.

**Pull request merged**